transhumandesign/kag-base
# [added] new component Flamer

## Status

- **READY**
Developed and tested in offline.
Tested briefly in online, no observed regressions or bugs.

## Description

Please note: PR #1374 has to be applied so the player can take damage even when fire hits are applied repeatedly in succession.

This PR adds a new interesting component tile "Flamer" that can be placed by a Builder in Sandbox.

- Requires 50 Stone to build.
- Has a stone background so as to not cause burning to itself indirectly.
- Will continuously spawn flame particles.
- Will burn wood blocks, grass or other flammable tiles.
- If the fire is emitting, will also emit a sound effect and a small light.
- Will ignite unlit fireplaces.
- To apply fire to other blobs, a radius check has been added. 
This seemed like the best approach to simulate a hurtbox.
A different approach that added a function to particles was neglected.
Within radius 7.5f of the tile it is facing, will ignite arrows.
Within radius 3.7f of the tile it is facing, will ignite unlit fireplaces, cook food and apply fire damage to blobs.
The applied fire damage only lasts 75% of the duration that the fire state normally does so touching Flamer fire is not too annoying.
- Will not emit fire for 30 ticks if hit by water splash.
- Will not emit fire if the exit is blocked or if the exit is underwater.

`ArrowCommon.as` has been added so Flamer can ignite Arrows.
`FireplaceCommon.as` has been added so Flamer can ignite Fireplaces.

Initially, a tick frequency of 3 was used but was changed to 1 to increase consistency of igniting Arrows.

I opted against the name being "Flamethrower" because Flamethrower sounds more like an item and Flamer in google images looks a lot like what I wanted the component to be like.
The sprite is a mockup by me. 
I don't remember where I got the fire sound, but it was most likely from here somewhere and edited by me: 
https://pixabay.com/sound-effects/search/fire/
If requested by the development team, I can change the sprite, sound or name.

## Steps to Test or Reproduce
Go to Sandbox and go to the devices tab, select Flamer and place it.
Place a source such as a lever, activate it.
Notice flames coming out.
Play around with it :-)

## Inspiration
There is not a direct inspiration but I kept thinking of these.
https://www.youtube.com/watch?v=mafCrNI6P70&t=25888s
https://www.youtube.com/watch?v=YAwXye_HlE4&t=365s
